cpp/src/arrow/util/hashing.h

```diff
@@ -687,12 +700,20 @@ class BinaryMemoTable : public MemoTable {
 
   // Same as above, but check output size in debug mode
   void CopyValues(int32_t start, int64_t out_size, uint8_t* out_data) const {
-    int32_t offset = offsets_[start];
-    auto length = values_.size() - static_cast<size_t>(offset);
+    if (start >= size()) {
+      return;
+    }
+
+    // The absolute byte offset of `start` value in the binary buffer.
+    int32_t offset = binary_builder_.OffsetOf(start);
+    auto length = binary_builder_.value_data_length() - static_cast<size_t>(offset);
+
     if (out_size != -1) {
       assert(static_cast<int64_t>(length) == out_size);
```

Comment:
Is it necessarily incorrect for length to be less than out_size?
```suggestion
      assert(static_cast<int64_t>(length) <= out_size);
```

Reply:
I think it should be left like this. It's a safeguard for the caller.